### PR TITLE
Add avatars and placeholders to the recipient autocomplete widget

### PR DESCRIPTION
**Pull request by ChristophWurst**
cc @jancborchardt @eppfel since we talked about that at the conf

would be great if a CSS ninja could fix the design a bit (e.g. some padding, fix vertical alignment and remove ugly background color/image). @skjnldsv maybe 😉 

bildschirmfoto von 2016-10-06 22-09-32

### CODE REVIEW

File: js/views/composerview.js [hunk 1]
```diff
@@ -468,7 +468,32 @@ define(function(require) {
...
+				}).data('ui-autocomplete')._renderItem = function ($ul, item) {
+					var $item = $('<li/>');
+					var $row = $('<a/>');
+
+					$row.addClass('mail-recipient-autocomplete');
+
+					if (item.photo && item.photo !== null) {
+						var $avatar = $('<img/>');
+						$avatar.addClass('avatar');
+						$avatar.height('32px');
```
Comment (ChristophWurst) [inline js/views/composerview.js]: we should move that to mail.css too

**Comment by MariusBluem** [comment]
> remove ugly background color/image

This would be also nice for the server in general 😁 I think the grey is a relict from very legacy versions 😜 @jancborchardt

**Comment by skjnldsv** [comment]
@MariusBluem It would be great to remove all background in general! I'm working on a night theme to prevent my eyes from dying (the white of nextcloud/owncloud is terrible! 😭) and there is so many white/grey background superposed this is becoming a nightmare to edit... 😞

**Comment by ChristophWurst** [comment]
> I'm working on a night theme to prevent my eyes from dying

Try http://jonls.dk/redshift/ 😎

**Comment by skjnldsv** [comment]
> Try http://jonls.dk/redshift/ 😎
I already am! But it's not enough. I'm usually forced to use my screen in negative! ^^

**Comment by skjnldsv** [comment]
@ChristophWurst Strange, avatars aren't displaying on my nc. Is it taking the data from contacts? I have avatars though! 😕

**Comment by ChristophWurst** [comment]
@skjnldsv do avatars work for you in the message list? There's currently a bug where it does not work on some setups. Cannot reproduce locally unfortunately.

What exactly do you see? Nothing? Yes, it takes your contacts images.

**Comment by MariusBluem** [comment]
> @MariusBluem It would be great to remove all background in general! I'm working on a night theme to prevent my eyes from dying (the white of nextcloud/owncloud is terrible! 😭) and there is so many white/grey background superposed this is becoming a nightmare to edit... 😞

I like this idea from a accessibility point of view 💪 Great!

**Comment by ChristophWurst** [comment]
> I see a list without images.

Could you inspect that row? Do you see any of the HTML that I've added with js?.

Did you enable debug mode *or* run the Makefile to minify the js files? Maybe you're using scripts from another branch

**Comment by jancborchardt** [comment]
Very nice @ChristophWurst! :) I’ll take it on if no one else wants to.

Some technical things:
- [ ] The apostrophes, be it single- or double-quotes, need to be removed. Otherwise the avatar placeholder displays them. And they are also just unnecessary.
- [ ] For contacts which have multiple email addresses, the email should be shown in half-grey behind or below the name. Otherwise they are impossible to distinguish.

**Comment by skjnldsv** [comment]
DOn't worry Christoph, the error was mine. Wasn't compiling properly.
Here you go: 
screen

**Comment by ChristophWurst** [comment]
> The apostrophes, be it single- or double-quotes, need to be removed. Otherwise the avatar placeholder displays them. And they are also just unnecessary.

IMO this is a separate issue of either contacts or collected addresses being stored wrongly. That should be fixed properly on the server side. We should refactor that anyways because we're now showing duplicates too.

File: css/mail.css [hunk 2]
```diff
@@ -389,6 +389,16 @@
 	right: -10px;
 }
 
+/* autocomplete list */
+.ui-autocomplete .ui-menu-item a.mail-recipient-autocomplete {
+	display: flex;
```
Comment (ChristophWurst) [inline css/mail.css]: nice!

**Comment by ChristophWurst** [comment]
> The apostrophes, be it single- or double-quotes, need to be removed. Otherwise the avatar placeholder displays them. And they are also just unnecessary.
> 
> IMO this is a separate issue of either contacts or collected addresses being stored wrongly. That should be fixed properly on the server side. We should refactor that anyways because we're now showing duplicates too.

Just checked: the ``oc_mail_collected_addresses`` contains mail address information in different formats. Most are stored as ``Jane Doe <jane@doe.de>``, but some like ``"Jane Doe" <jane@doe.de>"``. This could be fixed by a repair step.

**Comment by jancborchardt** [comment]
@ChristophWurst yeah, it sounds like a separate issue but actually in combination with this improvement it then looks strange with the placeholder avatars. ;) Like, you only see `"` in circles, not characters.

**Comment by skjnldsv** [comment]
Thanks :)
Btw, when multiple mail, I only got one displayed! :/

**Comment by ChristophWurst** [comment]
> For contacts which have multiple email addresses, the email should be shown in half-grey behind or below the name. Otherwise they are impossible to distinguish.

Brought back email addresses (there was an error in my code), but not yet half-grey. I'll first add the migration step to fix existing collected addresses and then we can do that visual enhancements for both contacts and collected addresses 🚀

**Comment by ChristophWurst** [comment]
> The apostrophes, be it single- or double-quotes, need to be removed. Otherwise the avatar placeholder displays them. And they are also just unnecessary.

As mentioned above, I'll submit a PR to fix that soon. A simple migration step should be able to sanitize the existing data and split it into recipient name and address (I'll add a new column to the database).

@jancborchardt @skjnldsv @Gomez please re-review this PR! Thanks

**Comment by skjnldsv** [comment]
LGTM 👍 

But I think we also should:
- Only display one avatar for the same contac
- Display the mail address and not only the name

**Comment by ChristophWurst** [comment]
> Only display one avatar for the same contac

Might be tricky since we now rely on the jQuery autocomplete to render the list.

> Display the mail address and not only the name

I fixed that with my last commit.

**Comment by ChristophWurst** [comment]
- [ ] TODO: fix unit tests

```
There was 1 failure:

1) ContactsIntegrationTest::testGetMatchingRecipient

Failed asserting that two arrays are equal.

--- Expected

+++ Actual

@@ @@

 Array (

     0 => Array (

         'id' => 1

         'label' => '"Jonathan Frakes" <jonathan@f...s.com>'

         'value' => '"Jonathan Frakes" <jonathan@f...s.com>'

+        'photo' => null

     )

     1 => Array (

         'id' => 2

         'label' => '"John Doe" <john@doe.info>'

         'value' => '"John Doe" <john@doe.info>'

+        'photo' => null

     )

     2 => Array (

         'id' => 2

         'label' => '"John Doe" <doe@john.info>'

         'value' => '"John Doe" <doe@john.info>'

+        'photo' => null

     )

 )
```

File: js/views/composerview.js [hunk 3]
```diff
@@ -474,7 +479,11 @@ define(function(require) {
 
 					$row.addClass('mail-recipient-autocomplete');
 
-					if (item.photo && item.photo !== null) {
+					if (prevUID == item['id']) {
```
Comment (ChristophWurst) [inline js/views/composerview.js]: use === instead please

File: js/views/composerview.js [hunk 4]
```diff
@@ -488,6 +497,8 @@ define(function(require) {
 						$row.append($placeholder);
 					}
 
+					prevUID = item['id'];
```
Comment (skjnldsv) [inline js/views/composerview.js]: Well, this is a requirement from the contact app and the vcard guidelines.
So I guess it should be? 
https://github.com/owncloud/contacts/blob/master/js/services/contact_service.js#L87

File: js/views/composerview.js (hunk 3 above)
Comment (skjnldsv) [inline js/views/composerview.js]: 👍

**Comment by jancborchardt** [comment]
@ChristophWurst did you see https://github.com/nextcloud/server/pull/1704 ;)

**Comment by ChristophWurst** [comment]
> @ChristophWurst did you see nextcloud/server#1704 ;)

yup :-)

**Comment by ChristophWurst** [comment]
> The apostrophes, be it single- or double-quotes, need to be removed. Otherwise the avatar placeholder displays them. And they are also just unnecessary.

Should be fixed by #151

**Comment by ChristophWurst** [comment]
👍 for @skjnldsv's part. Nice collaboration 😉

**Comment by skjnldsv** [comment]
Thanks! As always, a pleasure! ;)

**Pull request merged**